### PR TITLE
Piano Roll Editor Keyboard Zoom

### PR DESCRIPTION
**Pull request by blackears**
Allow the user to zoom the Piano Roll Editor window  using the keyboard.

### CODE REVIEW

File: mscore/pianoview.cpp [hunk 1]
```diff
@@ -499,10 +502,110 @@ int PianoView::pixelXToTick(int pixX) {
...
+    bool alt = event->modifiers() & Qt::AltModifier;
+
+      qDebug("Keypress: %d", event->key());
+
+      if (ctrl)
+      {
+            int cx = width() / 2;
+            int cy = height() / 2;
+
+            if (event->key() == Qt::Key_Minus) {
```
Comment (Spire42) [inline mscore/pianoview.cpp]: All four of these keyboard shortcuts should be user-configurable, not hard-coded.

Comment (blackears) [reply mscore/pianoview.cpp]: How would I do that?

File: mscore/pianoview.cpp [hunk 2]
```diff
@@ -499,10 +502,110 @@ int PianoView::pixelXToTick(int pixX) {
...
 
+
+//---------------------------------------------------------
+//   event
+//---------------------------------------------------------
+
+bool PianoView::event(QEvent *event)
+    {
+        if (event->type() == QEvent::KeyPress) {
+              QKeyEvent *ke = static_cast<QKeyEvent *>(event);
```
Comment (Howard-C) [inline mscore/pianoview.cpp]: Pointer declaration problem, please use `QKeyEvent* ke`.

File: mscore/pianoview.cpp [hunk 3]
```diff
@@ -499,10 +502,110 @@ int PianoView::pixelXToTick(int pixX) {
...
+
+bool PianoView::handleKeyPressEvent(QKeyEvent *event)
+    {
+    bool ctrl = event->modifiers() & Qt::ControlModifier;
+    bool alt = event->modifiers() & Qt::AltModifier;
+
+      qDebug("Keypress: %d", event->key());
+
+      if (ctrl)
+      {
```
Comment (Howard-C) [inline mscore/pianoview.cpp]: Brace belongs to the previous line.

File: mscore/pianoview.cpp [hunk 4]
```diff
@@ -499,10 +502,110 @@ int PianoView::pixelXToTick(int pixX) {
...
+
+
+//---------------------------------------------------------
+//   keyPressEvent
+//---------------------------------------------------------
+
+bool PianoView::handleKeyPressEvent(QKeyEvent *event)
+    {
+    bool ctrl = event->modifiers() & Qt::ControlModifier;
+    bool alt = event->modifiers() & Qt::AltModifier;
```
Comment (Howard-C) [inline mscore/pianoview.cpp]: 6-space indent for these three lines.

File: mscore/pianoview.cpp [hunk 5]
```diff
@@ -499,10 +502,110 @@ int PianoView::pixelXToTick(int pixX) {
...
+      {
+            int cx = width() / 2;
+            int cy = height() / 2;
+
+            if (event->key() == Qt::Key_Minus) {
+                  zoomView(-1, !alt, cx, cy);
+                  return true;
+                }
+            if (event->key() == Qt::Key_Equal) {
+                zoomView(1, !alt, cx, cy);
```
Comment (Howard-C) [inline mscore/pianoview.cpp]: 6-space indent too.

File: mscore/pianoview.cpp [hunk 6]
```diff
@@ -499,10 +502,110 @@ int PianoView::pixelXToTick(int pixX) {
...
+
+//---------------------------------------------------------
+//   zoomView
+//---------------------------------------------------------
+
+void PianoView::zoomView(int step, bool horizontal, int centerX, int centerY)
+    {
+    if (horizontal) {
+        //Horizontal zoom
+        QRectF viewRect = mapToScene(viewport()->geometry()).boundingRect();
```
Comment (Howard-C) [inline mscore/pianoview.cpp]: 6-space indent here too.

File: mscore/pianoview.h [hunk 7]
```diff
@@ -116,13 +116,16 @@ class PianoView : public QGraphicsView {
       QAction* getAction(const char* id);
 
    protected:
+      virtual bool event(QEvent *event);
```
Comment (Howard-C) [inline mscore/pianoview.h]: Pointer declaration problem here too.

File: mscore/pianoview.h [hunk 8]
```diff
@@ -116,13 +116,16 @@ class PianoView : public QGraphicsView {
...
       virtual void wheelEvent(QWheelEvent* event);
       virtual void mousePressEvent(QMouseEvent* event);
       virtual void mouseReleaseEvent(QMouseEvent* event);
       virtual void mouseMoveEvent(QMouseEvent* event);
       virtual void leaveEvent(QEvent*);
       virtual void contextMenuEvent(QContextMenuEvent *event);
 
+      bool handleKeyPressEvent(QKeyEvent *event);
```
Comment (Howard-C) [inline mscore/pianoview.h]: Here too.

**Comment by Jojo-Schmitz** [comment]
Why share/sound/MuseScore_General-License.md  and share/sound/MuseScore_General-changelog.txt ?

File: mscore/shortcut.cpp [hunk 9]
```diff
@@ -3741,6 +3741,43 @@ Shortcut Shortcut::_sc[] = {
          Icons::mail_ICON,
          Qt::ApplicationShortcut
          },
+      {
+         MsWidget::PIANO_ROLL_EDITOR,
+         STATE_ALL,
+         "zoom-in-horiz-pre",
+         QT_TRANSLATE_NOOP("action", "Zoom In Horizontally"),
+         QT_TRANSLATE_NOOP("action", "Zoom In Horizontal - Piano Roll Editor"),
```
Comment (Jojo-Schmitz) [inline mscore/shortcut.cpp]: Here we'd want Sentance case, not Title Case

File: mscore/shortcut.cpp [hunk 10]
```diff
@@ -3741,6 +3741,43 @@ Shortcut Shortcut::_sc[] = {
...
+         0,
+         Icons::Invalid_ICON
+         },
+      {
+         MsWidget::PIANO_ROLL_EDITOR,
+         STATE_ALL,
+         "zoom-out-horiz-pre",
+         QT_TRANSLATE_NOOP("action", "Zoom Out Horizontally"),
+         QT_TRANSLATE_NOOP("action", "Zoom Out Horizontal - Piano Roll Editor"),
```
Comment (Jojo-Schmitz) [inline mscore/shortcut.cpp]: here too

File: mscore/shortcut.cpp [hunk 11]
```diff
@@ -3741,6 +3741,43 @@ Shortcut Shortcut::_sc[] = {
...
+         0,
+         Icons::Invalid_ICON
+         },
+      {
+         MsWidget::PIANO_ROLL_EDITOR,
+         STATE_ALL,
+         "zoom-in-vert-pre",
+         QT_TRANSLATE_NOOP("action", "Zoom In Vertically"),
+         QT_TRANSLATE_NOOP("action", "Zoom In Vertical - Piano Roll Editor"),
```
Comment (Jojo-Schmitz) [inline mscore/shortcut.cpp]: here too

File: mscore/shortcut.cpp [hunk 12]
```diff
@@ -3741,6 +3741,43 @@ Shortcut Shortcut::_sc[] = {
...
+         0,
+         Icons::Invalid_ICON
+         },
+      {
+         MsWidget::PIANO_ROLL_EDITOR,
+         STATE_ALL,
+         "zoom-out-vert-pre",
+         QT_TRANSLATE_NOOP("action", "Zoom Out Vertically"),
+         QT_TRANSLATE_NOOP("action", "Zoom Out Vertical - Piano Roll Editor"),
```
Comment (Jojo-Schmitz) [inline mscore/shortcut.cpp]: here too

File: mscore/pianoroll.cpp [hunk 13]
```diff
@@ -816,4 +845,19 @@ void PianorollEditor::tickLenChanged(int val)
...
+      int cx = pianoView->width() / 2;
+      int cy = pianoView->height() / 2;
+
+      pianoView->zoomView(amount, horiz, cx, cy);
+      }
+
 }
+
+
+
```
Comment (Jojo-Schmitz) [inline mscore/pianoroll.cpp]: why those 3 extra empty lines?

**Comment by Jojo-Schmitz** [comment]
no such shortcuts for macOS and/or AZERTY?

**Comment by RobFog** [comment]
Rebase needed.

**Pull request merged**